### PR TITLE
[V1] `AsyncLLM` Implementation

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
### SUMMARY:
* `AsyncLLM` in V1 - better overlapping of GPU and CPU

### TODO:
- [x] Make mypy happy
- [x] Remove debugging polling in `io threads`
- [x] LM Eval testing
- [x] Cleaner shutdown
- [x] Get the test passing in the CI :)

### FOLLOW UP PRS:
* Benchmarking with CUDAGraphs (todo as follow up given cudagraphs are broken)
* Robustness (health checks, make sure abort is working properly everywhere)
* More `AsyncLLM` and `LLMEngine` tests (abort, stop string, other unit)
* Enable multiprocessing for `LLM` by default (need to figure out a way around fork) - currently, need to set `VLLM_ENABLE_V1_MULTIPROCESSING=1`

### DIAGRAM:
* Note: this diagram is a bit dated. There is an `EngineCoreClient` class that is used by the `AsyncLLM` to interact with the `EngineCore`, but the overall architecture is close to what we have.
* Note: stop strings are detected in the detokenizer and we send an abort message from `output_handler_loop` to `EngineCore`

image

### CODE REVIEW

File: vllm/v1/engine/async_llm_engine.py
Comment (robertgshaw2-neuralmagic) [inline vllm/v1/engine/async_llm_engine.py:40]: TODO: update to avoid passing this full long list everywhere

File: vllm/v1/engine/detokenizer.py
Comment (robertgshaw2-neuralmagic) [inline vllm/v1/engine/detokenizer.py:37]: TODO: remove stream mode to simplify for now

File: vllm/v1/core/scheduler.py
Comment (robertgshaw2-neuralmagic) [inline vllm/v1/core/scheduler.py:231]: Im not sure it makes sense for this method to be in `scheduler.py`

The only item related to setting the scheduler here is updating which `self.running`

File: vllm/entrypoints/llm.py
Comment (robertgshaw2-neuralmagic) [inline vllm/entrypoints/llm.py:205]: @njhill - im starting to think this stuff should be wrapped into `llmengine`

Comment (njhill) [reply vllm/entrypoints/llm.py:205]: @robertgshaw2-neuralmagic yeah I was thinking we might have a completely separate `LLM` class but that may be tricky if we want to be able to switch existing code with the env var.

File: vllm/v1/core/scheduler.py
Comment (robertgshaw2-neuralmagic) [inline vllm/v1/core/scheduler.py:231]: @njhill do you think we should remove this method and instead have this implemented outside EngineCore (and then sending `abort` when stopping criteria?)

This might cause extra work. But it would enable fully async scheduling since the current inner loop is:

```python
scheduler.schedule()
executor.execute()
scheduler.update_from_output()
```

Comment (robertgshaw2-neuralmagic) [reply vllm/v1/core/scheduler.py:231]: cc @WoosukKwon

File: vllm/v1/core/scheduler.py
Comment (njhill) [inline vllm/v1/core/scheduler.py:65]: I don't think this is used anywhere?

File: vllm/v1/engine/__init__.py
Comment (njhill) [inline vllm/v1/engine/__init__.py:28]: Can we add `array_like=True` here too?

File: vllm/v1/engine/__init__.py
Comment (njhill) [inline vllm/v1/engine/__init__.py:69]: These should probably be upper camel-case per convention

File: vllm/v1/engine/core.py
Comment (njhill) [inline vllm/v1/engine/core.py:293]: curious what using a timeout here buys us?

File: vllm/v1/engine/core.py
Comment (njhill) [inline vllm/v1/engine/core.py:333]: Similar question, why do we need to poll with timeout here?

File: vllm/v1/engine/core.py
Comment (njhill) [inline vllm/v1/engine/core.py:338]: nit:
```suggestion
                type_frame, data_frame = socket.recv_multipart(copy=False)
                request_type = type_frame.buffer
                request_data = data_frame.buffer
```

File: vllm/v1/engine/core.py
Comment (njhill) [inline vllm/v1/engine/core.py:361]: same question :)

**Review by njhill — COMMENTED** [review]
I did a quick pass, it's looking great thanks @robertgshaw2-neuralmagic 

I'll continue to review but I don't see why shouldn't merge it soon and continue to iterate with more PRs

File: vllm/v1/engine/core.py
Comment (robertgshaw2-neuralmagic) [reply vllm/v1/engine/core.py:293]: I was using it for debugging.

File: vllm/v1/engine/core.py
Comment (robertgshaw2-neuralmagic) [reply vllm/v1/engine/core.py:333]: Debugging

File: vllm/v1/engine/core.py
Comment (robertgshaw2-neuralmagic) [reply vllm/v1/engine/core.py:361]: debugging

File: vllm/v1/core/scheduler.py
Comment (robertgshaw2-neuralmagic) [inline vllm/v1/core/scheduler.py:258]: note: this could be done elsewhere at the cost of another loop

File: vllm/v1/engine/__init__.py
Comment (robertgshaw2-neuralmagic) [reply vllm/v1/engine/__init__.py:28]: I think its better to not use `array_like` here since we get better logs without it and these are sent infrequently

File: vllm/engine/output_processor/stop_checker.py
Comment (robertgshaw2-neuralmagic) [inline vllm/engine/output_processor/stop_checker.py:92]: @njhill - can you add unit tests for this?

File: vllm/v1/engine/core.py
Comment (robertgshaw2-neuralmagic) [reply vllm/v1/engine/core.py:295]: actually nvm, this task will be canceled by shutdown

Comment (robertgshaw2-neuralmagic) [reply vllm/v1/engine/core.py:295]: Since this runs in the main thread, we need to be careful about shutdown

Comment (robertgshaw2-neuralmagic) [reply vllm/v1/engine/core.py:295]: whereas the io handlers are `daemon`

File: vllm/v1/engine/async_llm.py
Comment (njhill) [inline vllm/v1/engine/async_llm.py:288]: This we should definitely eliminate imo

File: vllm/engine/output_processor/stop_checker.py
Comment (njhill) [reply vllm/engine/output_processor/stop_checker.py:92]: It's covered by https://github.com/vllm-project/vllm/blob/main/tests/engine/test_stop_strings.py, but we could add some more isolated tests too

File: vllm/v1/engine/core.py
Comment (robertgshaw2-neuralmagic) [reply vllm/v1/engine/core.py:271]: Basically I think that any error in `step()` is unrecoverable since `step()` operates on every `RUNNING` request in the system simultaneously. Usually the issue is some NCCL or CUDA error which are unrecoverable.

That being said, there is a class of cases where the failures are "request level" --- these usually occur due to a bad logits processor that can be specified by the user or some other DoS bug (e.g. we have had cases where sending a token out of vocabulary crashes the whole system).

I think we should try to detect these long before they get into `EngineCore`. Since each individual request is now sent to the engine core from an separate asyncio task, this is a very natural place since any errors will just crash that since asyncio task.

I wish there were a way to recover from a failed `step()` though. The follow up from this PR is to propogate this issue back to `AsyncLLM` (via the equivalent of `AsyncEngineDeadError` that we have on V0). 

I think recovering from `step()` failures is something we could consider in the future, but its out of the scope of the PR.

File: vllm/v1/engine/async_llm.py
Comment (robertgshaw2-neuralmagic) [reply vllm/v1/engine/async_llm.py:338]: These will be eliminated in the end. For now, we need to conform to `EngineClient` so we dont have to change the server API

File: vllm/v1/engine/core.py
Comment (russellb) [reply vllm/v1/engine/core.py:278]: Thank you for the detailed response! This is helpful!

**Review by njhill — APPROVED** [review]
Thanks @robertgshaw2-neuralmagic, I think this is in good shape to land and continue to improve via other updates.

File: vllm/v1/engine/__init__.py
Comment (tlrmchlsmth) [inline vllm/v1/engine/__init__.py:1]: nit: might be nice to have this stuff in a file with a meaningful name like `core_io.py` or `engine_ipc.py` or something better than that, since these are all things that are serialized, shoved into message queues, and sent between processes

File: vllm/v1/engine/async_llm.py
Comment (tlrmchlsmth) [inline vllm/v1/engine/async_llm.py:305]: try to delete before landing?

File: vllm/outputs.py
Comment (mgoin) [inline vllm/outputs.py:144]: Is there something we could assert on if logprobs aren't implemented but requested?

File: vllm/v1/engine/core.py
Comment (tlrmchlsmth) [inline vllm/v1/engine/core.py:143]: With #9856, we'll have separate worker processes running the forward pass, so we can't rely on the queue-consumer thread calling into CUDA kernels to release the GIL in that case. 

I think this should be fine, since in that case we call `os.sched_yield()` - I'll look into that. But we might want to make a note in this comment that the consumer thread must release the GIL one way or another

File: vllm/v1/engine/async_llm.py
Comment (mgoin) [inline vllm/v1/engine/async_llm.py:123]: Should you leave a TODO for other executors to be supported?

File: vllm/v1/engine/async_llm.py
Comment (mgoin) [inline vllm/v1/engine/async_llm.py:305]: I think this is used in streaming cases where the client decides they don't want the request to finish. I personally cancel requests early when the output is clearly not what I want, so I think it is a valid usecase to eventually implement

File: vllm/v1/engine/core.py
Comment (mgoin) [inline vllm/v1/engine/core.py:41]: What is the FIXME for?

**Review by mgoin — APPROVED** [review]
Amazing work and results folks, leaving comments but they are not blocking

File: vllm/v1/engine/__init__.py
Comment (tlrmchlsmth) [inline vllm/v1/engine/__init__.py:25]: Any reason to specify `array_like=True` and `gc=False` for the other cases but not this one? 

BTW did you see any performance improvement for `array_like=True`? We might want to make it easy to switch all of these over to `array_like=False` because
> This may improve performance, at the cost of a more inscrutable message encoding.

(from https://jcristharif.com/msgspec/api.html)

Not sure how inscrutable they get though

File: vllm/v1/engine/async_llm.py
Comment (njhill) [reply vllm/v1/engine/async_llm.py:305]: Cancellation is done by closing/discarding the generator that the generate() method returns (used by client / front-end to stream the results). That already works.

This `abort` method is an additional "out of band" way of canceling, the question is whether we need it too.

**Review by tlrmchlsmth — APPROVED** [review]
Great work - left some comments but all minor stuff + notes for future improvements

File: vllm/outputs.py
Comment (robertgshaw2-neuralmagic) [reply vllm/outputs.py:144]: This is the same as the current state of V1.

Feldman has PR to add logprobs which will land quickly so I think this is fine to leave as is

File: vllm/v1/engine/__init__.py
Comment (robertgshaw2-neuralmagic) [reply vllm/v1/engine/__init__.py:25]: Mostly that `EngineCoreRequests` are infrequent in comparison to `EngineCoreOutput`. 

So we wanted to have the error message encoding since it does not really matter for performance

Comment (robertgshaw2-neuralmagic) [reply vllm/v1/engine/__init__.py:25]: I measured some benefits from array like. Not monumental though. I think we can do a separate experiment as a follow up PR

File: vllm/v1/engine/async_llm.py
Comment (robertgshaw2-neuralmagic) [reply vllm/v1/engine/async_llm.py:123]: This is consistent with the current state of V1. This is a big task to support other Executors so I am not worried about this falling through the cracks

File: vllm/v1/engine/async_llm.py
Comment (robertgshaw2-neuralmagic) [reply vllm/v1/engine/async_llm.py:305]: We need this to conform to the `EngineClient` ABC. See nick's point below re: why this is probably not needed (but needs changes to vllm that are outside the scope of this PR). I think this needs to be kept for unit testing.

File: vllm/v1/engine/core.py
Comment (njhill) [reply vllm/v1/engine/core.py:143]: @tlrmchlsmth I've been thinking a lot about this and was just looking at your PR (which looks great), would be good to talk though some ideas soon when you have a chance.

File: vllm/v1/engine/__init__.py
Comment (njhill) [reply vllm/v1/engine/__init__.py:25]: Where do we actually care about/log the wire-encoded contents though? This is for internal IPC so it shouldn't matter imo, better to be compact. It's like protobuf, I'm not sure how readable the msgpack encoding is anyhow.

That said it shouldn't be hard to make this a toggle as suggested.

File: vllm/v1/engine/async_llm.py
Comment (mgoin) [reply vllm/v1/engine/async_llm.py:305]: I see the difference now, thanks

File: tests/entrypoints/llm/test_accuracy.py
Comment (joerunde) [reply tests/entrypoints/llm/test_accuracy.py:57]: I agree!

I can also go simplify tests post-merge if needed

File: vllm/entrypoints/openai/api_server.py
Comment (joerunde) [inline vllm/entrypoints/openai/api_server.py:65]: This won't work with tests, anything that tries to monkeypatch with `m.setenv("VLLM_USE_V1", True)` won't take effect because this runs the check at import time

**Comment by joerunde** [comment]
@robertgshaw2-neuralmagic @njhill this looks super 🚀🚀🚀 

Couple questions for my own understanding:
- How should I interpret the name `core`? There's both a `v1/core` package as well as `v1/engine/core.py`
- It looks like this is way faster than v0 + multistep decoding, are we planning on ditching multistep in v1 or is that still TBD?

Thanks!

File: tests/v1/engine/test_async_llm.py
Comment (joerunde) [inline tests/v1/engine/test_async_llm.py:38]: Definitely out of scope for this PR but regarding the followup point

> More AsyncLLM and LLMEngine tests (abort, stop string, other unit)

We should be able to reuse all the existing tests since the interfaces are the same, right? It'd just be a matter of hooking up a fixture to set the appropriate `v1` environment variables and making sure we're initializing the engine under test with a method that returns the appropriate one. I'd be happy to make that my project for the next few days while y'all focus on building the fast stuff

Comment (joerunde) [reply tests/v1/engine/test_async_llm.py:38]: Ah, yeah of course. There's a `@skip_v1` mark for tests for unsupported features.

> some of the internal classes that are used for testing (e.g. scheduled) are now in engine core, so we cannot access them directly from llm_engine

Oof, yeah those sound like some unideal nosy tests, I'll have to look more at them

File: vllm/envs.py
Comment (WoosukKwon) [inline vllm/envs.py:480]: QQ: In which case should we turn this on?

**Review by WoosukKwon — APPROVED** [review]
Thanks for the great work!

File: vllm/envs.py
Comment (robertgshaw2-neuralmagic) [reply vllm/envs.py:480]: `VLLM_ENABLE_V1_MULTIPROCESSING=1` enables multiprocessing for `EngineCore` inside `LLM` (multiprocessing is used for `AsyncLLM`). It is faster than the current implementation.

image

We will want to enable `VLLM_ENABLE_V1_MULTIPROCESSING=1`, but right now it is a problem for `LLM` since we cannot `spawn` without an `if __name__ == "__main__"` guard. We left solving this issue for follow up work.

File: vllm/v1/engine/__init__.py
Comment (robertgshaw2-neuralmagic) [reply vllm/v1/engine/__init__.py:1]: will do in follow up

**Comment by robertgshaw2-neuralmagic** [comment]
Hey @lixiaolx - thanks for taking a look. I am having a hard time understanding your analysis - could you clarify?

**Comment by njhill** [comment]
Thanks @lixiaolx, nice profiles! What you observe is not unexpected since the scheduling logic currently contends for the GIL with the IPC message serialization/deserialization.

Our intention is to improve this very soon but doing the IPC work in a separate thread is still a big win as a first step since much of that work overlaps with parts of the critical loop that _don't_ contend for the GIL, primarily the forward pass in the GPU.

**Comment by lixiaolx** [comment]
> Thanks @lixiaolx, nice profiles! What you observe is not unexpected since the scheduling logic currently contends for the GIL with the IPC message serialization/deserialization.
> 
> Our intention is to improve this very soon but doing the IPC work in a separate thread is still a big win as a first step since much of that work overlaps with parts of the critical loop that _don't_ contend for the GIL, primarily the forward pass in the GPU.

Thank you very much for your answer. I tried to compare this solution. If we solve the GIL problem, the remaining gap time will be 2-3ms according to the above calculation. 
I would like to ask if we have any plans to do asynchronous scheduling? Compared with sglang asynchronous, there is still a gap. 
I recently analyzed that the overall test gap of sglang's asynchronous solution under the same conditions is between 200-300us. If you have a plan, are there any arrangements?

**Comment by lixiaolx** [comment]
> Hey @lixiaolx - thanks for taking a look. I am having a hard time understanding your analysis - could you clarify?
@robertgshaw2-neuralmagic 
I compared the previous pr with your current pr, and did nsys analysis. I added nvtx to analyze the time overhead where the mainloop function is called, and split and analyzed the CPU overhead between the two forwards before and after the GPU.

**Comment by lixiaolx** [comment]
@robertgshaw2-neuralmagic @njhill   Hello, does our pr support multiple gpu cards? Well, when testing llama2-70b 8gpu，occurs server log  was stuck here.
image
I use nvidia-smi found that only 0 gpu card was occupying only about 500MB.

**Comment by njhill** [comment]
@lixiaolx the V1 path is still in an alpha state and does not yet support multiple GPUs, but will do soon.

**Comment by lixiaolx** [comment]
> > Thanks @lixiaolx, nice profiles! What you observe is not unexpected since the scheduling logic currently contends for the GIL with the IPC message serialization/deserialization.
> > Our intention is to improve this very soon but doing the IPC work in a separate thread is still a big win as a first step since much of that work overlaps with parts of the critical loop that _don't_ contend for the GIL, primarily the forward pass in the GPU.
> 
> Thank you very much for your answer. I tried to compare this solution. If we solve the GIL problem, the remaining gap time will be 2-3ms according to the above calculation. I would like to ask if we have any plans to do asynchronous scheduling? Compared with sglang asynchronous, there is still a gap. I recently analyzed that the overall test gap of sglang's asynchronous solution under the same conditions is between 200-300us. If you have a plan, are there any arrangements?

@njhill ,Is there any arrangement for this asynchronous scheduling?

**Comment by lixiaolx** [comment]
> @lixiaolx the V1 path is still in an alpha state and does not yet support multiple GPUs, but will do soon.

OK，thank you

**Comment by njhill** [comment]
> @njhill ,Is there any arrangement for this asynchronous scheduling?

Not yet, our plan is to first optimize other aspects first since it will be complex to combine this with certain other optimizations.

**Comment by B-201** [comment]
> > Hey @lixiaolx - thanks for taking a look. I am having a hard time understanding your analysis - could you clarify?
> > @robertgshaw2-neuralmagic
> > I compared the previous pr with your current pr, and did nsys analysis. I added nvtx to analyze the time overhead where the mainloop function is called, and split and analyzed the CPU overhead between the two forwards before and after the GPU.

Sorry to bother you, but I’d like to ask how you added nvtx to analyze the time overhead of these function calls?

**Pull request merged**
